namenode/src/main/java/org/apache/crail/namenode/BlockStore.java

```diff
@@ -171,9 +192,33 @@ short addDataNode(DataNodeBlocks dataNode) {
 		return RpcErrors.ERR_OK;
 
 	}
-	
+
+	short prepareForRemovalDatanode(DataNodeInfo dn) throws Exception {
+		// this will only mark it for removal
+		return prepareOrRemoveDN(dn, true);
+	}
+
+	short removeDatanode(DataNodeInfo dn) throws Exception {
+		// this will remove it as well
+		return prepareOrRemoveDN(dn, false);
+	}
+
 	//---------------
-	
+
+	private short prepareOrRemoveDN(DataNodeInfo dn, boolean onlyMark) throws Exception {
+		DataNodeBlocks toBeRemoved = membership.get(dn.key());
+		if (toBeRemoved == null) {
+			LOG.error("DataNode: " + dn.toString() + " not found");
+			return RpcErrors.ERR_DATANODE_NOT_REGISTERED;
+		} else {
+			if (onlyMark)
+				toBeRemoved.scheduleForRemoval();
+			else
+				membership.remove(toBeRemoved.key());
```

Comment:
Can you explain what happens when we do not mark but just remove?

Reply:
A datanode is only marked for removal first because it might still store datablocks. Only when no data is stored anymore by a marked datanode the namenode can start the actual removal process, that is removing it from its datastructures and notifying the datanode that it can terminate (this is periodically checked as part of the `getDataNode` RPC). When we just remove that datanode the data it stores is not available anymore.

Reply:
So in the case the datanode still has data and is forcefully removed, we would get file with wholes in them, correct? Blocks of this datanodes that are still used in files are not updated that the datanode has been removed? So if a client asks for a block he will get the datanode info of a non-existing datanode? Worse if someone starts a datanode with the same ip/port at a later point he might see different data (depending on datanode type).

Reply:
I'm not sure I fully understand the problem you are referring to. Could you rephrase perhaps?
My understanding is that when a datanode is forcefully removed (also independently of this PR, e.g. by killing it) we run into the problem you mentioned that certain blocks of files might be missing. When a client requests a file the namenode will point to the corresponding datanodes. As the datanode was forcefully shutdown the client application will not be able to retrieve the missing blocks. Am I missing a point here?
To prevent this situation the idea was to only allow the (actual) removal of a datnode when it does not store any remaining blocks.

Reply:
Sorry for the confusion I was not referring to killing a datanode. Is this method always called first with `onlyMark=true` (e.g. prepare) and later you remove or is there a case where one can "force" remove and call with `onlyMark=false` directly?
I think given that this method does two different things we should split it in a prepare and a remove method.

Reply:
Thank you for clarifying. Yes, that is correct a datanode is always marked first and removed afterwards when there are no remaining blocks (i.e. there is no "force" case).
I agree splitting this into two methods would make things clearer and avoid possible confusions. I will change it.

Reply:
Thanks!

Reply:
I hope the new organization of the methods makes more sense.